// @fast-check/jest/lib/cjs/jest-fast-check-worker.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_TestBuilder = require("./TestBuilder-ocZZrPwv.js");
let fast_check = require("fast-check");
fast_check = require_TestBuilder.__toESM(fast_check);
let _fast_check_worker = require("@fast-check/worker");
let _jest_expect = require("@jest/expect");
//#region src/jest-fast-check-worker.ts
function typedAssign(fun, others) {
	return Object.assign(fun, others);
}
function dummyTest() {
	return typedAssign(() => void 0, {
		concurrent: typedAssign(() => void 0, {
			each: () => () => void 0,
			failing: typedAssign(() => void 0, { each: () => () => void 0 }),
			only: typedAssign(() => void 0, {
				each: () => () => void 0,
				failing: typedAssign(() => void 0, { each: () => () => void 0 })
			}),
			skip: typedAssign(() => void 0, {
				each: () => () => void 0,
				failing: typedAssign(() => void 0, { each: () => () => void 0 })
			})
		}),
		each: () => () => void 0,
		failing: typedAssign(() => void 0, { each: () => () => void 0 }),
		only: typedAssign(() => void 0, {
			each: () => () => void 0,
			failing: typedAssign(() => void 0, { each: () => () => void 0 })
		}),
		skip: typedAssign(() => void 0, {
			each: () => () => void 0,
			failing: typedAssign(() => void 0, { each: () => () => void 0 })
		}),
		todo: () => void 0
	});
}
const init = (url, options) => {
	const fcExtra = {
		asyncProperty: (0, _fast_check_worker.propertyFor)(url, options),
		assert: _fast_check_worker.assert,
		readConfigureGlobal: fast_check.readConfigureGlobal
	};
	if (typeof it !== "undefined") if (typeof jest !== "undefined") return {
		test: require_TestBuilder.buildTest(test, jest, fcExtra),
		it: require_TestBuilder.buildTest(it, jest, fcExtra),
		expect: _jest_expect.jestExpect
	};
	else return import("@jest/globals").then(({ jest }) => ({
		test: require_TestBuilder.buildTest(test, jest, fcExtra),
		it: require_TestBuilder.buildTest(it, jest, fcExtra),
		expect: _jest_expect.jestExpect
	}));
	const dummyJest = {};
	return {
		test: require_TestBuilder.buildTest(dummyTest(), dummyJest, fcExtra),
		it: require_TestBuilder.buildTest(dummyTest(), dummyJest, fcExtra),
		expect: _jest_expect.jestExpect
	};
};
//#endregion
Object.defineProperty(exports, "fc", {
	enumerable: true,
	get: function() {
		return fast_check;
	}
});
exports.init = init;
