// @fast-check/jest/lib/internals/TestPropBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTestProp = void 0;
const fast_check_1 = require("fast-check");
const globals_1 = require("@jest/globals");
const TestWithPropRunnerBuilder_js_1 = require("./TestWithPropRunnerBuilder.js");
function internalTestPropFailing(testFn) {
    function base(label, arbitraries, prop, params) {
        (0, TestWithPropRunnerBuilder_js_1.buildTestWithPropRunner)(testFn, label, arbitraries, prop, params, undefined, globals_1.jest, fast_check_1.default);
    }
    const extras = {};
    return Object.assign(base, extras);
}
function internalTestPropBase(testFn) {
    function base(label, arbitraries, prop, params) {
        (0, TestWithPropRunnerBuilder_js_1.buildTestWithPropRunner)(testFn, label, arbitraries, prop, params, undefined, globals_1.jest, fast_check_1.default);
    }
    const extras = {
        failing: internalTestPropFailing(testFn.failing),
    };
    return Object.assign(base, extras);
}
function internalTestPropConcurrent(testFn) {
    function base(label, arbitraries, prop, params) {
        (0, TestWithPropRunnerBuilder_js_1.buildTestWithPropRunner)(testFn, label, arbitraries, prop, params, undefined, globals_1.jest, fast_check_1.default);
    }
    const extras = {
        only: internalTestPropBase(testFn.only),
        skip: internalTestPropBase(testFn.skip),
        failing: internalTestPropFailing(testFn.failing),
    };
    return Object.assign(base, extras);
}
function internalTestProp(testFn) {
    const base = internalTestPropConcurrent(testFn);
    const extras = {
        concurrent: internalTestPropConcurrent(testFn.concurrent),
        todo: testFn.todo,
    };
    return Object.assign(base, extras);
}
exports.buildTestProp = internalTestProp;
