// @yarnpkg/sdks/lib/sdks/vscode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VSCODE_SDKS = exports.generateFlowBinWrapper = exports.generateSvelteLanguageServerWrapper = exports.generateTypescriptWrapper = exports.generateRelayCompilerWrapper = exports.generatePrettierWrapper = exports.generateEslintWrapper = exports.generateAstroLanguageServerWrapper = exports.generateDefaultWrapper = exports.addVSCodeWorkspaceConfiguration = exports.VSCodeConfiguration = void 0;
const tslib_1 = require("tslib");
const fslib_1 = require("@yarnpkg/fslib");
const sdkUtils = tslib_1.__importStar(require("../sdkUtils"));
var VSCodeConfiguration;
(function (VSCodeConfiguration) {
    VSCodeConfiguration["settings"] = "settings.json";
    VSCodeConfiguration["extensions"] = "extensions.json";
})(VSCodeConfiguration || (exports.VSCodeConfiguration = VSCodeConfiguration = {}));
const addVSCodeWorkspaceConfiguration = async (pnpApi, type, patch) => {
    const relativeFilePath = `.vscode/${type}`;
    await sdkUtils.addSettingWorkspaceConfiguration(pnpApi, relativeFilePath, patch);
};
exports.addVSCodeWorkspaceConfiguration = addVSCodeWorkspaceConfiguration;
const generateDefaultWrapper = async (pnpApi) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`search.exclude`]: {
            [`**/.yarn`]: true,
            [`**/.pnp.*`]: true,
        },
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `arcanis.vscode-zipfs`,
        ],
    });
};
exports.generateDefaultWrapper = generateDefaultWrapper;
const generateAstroLanguageServerWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`astro.language-server.ls-path`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`bin/nodeServer.js`)),
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `astro-build.astro-vscode`,
        ],
    });
};
exports.generateAstroLanguageServerWrapper = generateAstroLanguageServerWrapper;
const generateEslintWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`eslint.nodePath`]: fslib_1.npath.fromPortablePath(fslib_1.ppath.dirname(fslib_1.ppath.dirname(wrapper.getProjectPathTo(fslib_1.Filename.manifest)))),
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `dbaeumer.vscode-eslint`,
        ],
    });
};
exports.generateEslintWrapper = generateEslintWrapper;
const generatePrettierWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`prettier.prettierPath`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(fslib_1.ppath.normalize(wrapper.manifest.main))),
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `esbenp.prettier-vscode`,
        ],
    });
};
exports.generatePrettierWrapper = generatePrettierWrapper;
const generateRelayCompilerWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`relay.pathToRelay`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`cli.js`)),
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `meta.relay`,
        ],
    });
};
exports.generateRelayCompilerWrapper = generateRelayCompilerWrapper;
const generateTypescriptWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`typescript.tsdk`]: fslib_1.npath.fromPortablePath(fslib_1.ppath.dirname(wrapper.getProjectPathTo(`lib/tsserver.js`))),
        [`typescript.enablePromptUseWorkspaceTsdk`]: true,
    });
};
exports.generateTypescriptWrapper = generateTypescriptWrapper;
const generateSvelteLanguageServerWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`svelte.language-server.ls-path`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`bin/server.js`)),
        [`svelte.language-server.runtime-args`]: [`--loader`, `./.pnp.loader.mjs`],
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `svelte.svelte-vscode`,
        ],
    });
};
exports.generateSvelteLanguageServerWrapper = generateSvelteLanguageServerWrapper;
const generateFlowBinWrapper = async (pnpApi, target, wrapper) => {
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.settings, {
        [`flow.pathToFlow`]: fslib_1.npath.fromPortablePath(`\${workspaceFolder}/${wrapper.getProjectPathTo(`cli.js`)}`),
    });
    await (0, exports.addVSCodeWorkspaceConfiguration)(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `flowtype.flow-for-vscode`,
        ],
    });
};
exports.generateFlowBinWrapper = generateFlowBinWrapper;
exports.VSCODE_SDKS = [
    [null, exports.generateDefaultWrapper],
    [`@astrojs/language-server`, exports.generateAstroLanguageServerWrapper],
    [`eslint`, exports.generateEslintWrapper],
    [`prettier`, exports.generatePrettierWrapper],
    [`relay-compiler`, exports.generateRelayCompilerWrapper],
    [`typescript-language-server`, null],
    [`typescript`, exports.generateTypescriptWrapper],
    [`svelte-language-server`, exports.generateSvelteLanguageServerWrapper],
    [`flow-bin`, exports.generateFlowBinWrapper],
];
