// @fast-check/jest/lib/esm/internals/TestPropBuilder.js
import fc from 'fast-check';
import { jest } from '@jest/globals';
import { buildTestWithPropRunner } from './TestWithPropRunnerBuilder.js';
function internalTestPropFailing(testFn) {
    function base(label, arbitraries, prop, params) {
        buildTestWithPropRunner(testFn, label, arbitraries, prop, params, undefined, jest, fc);
    }
    const extras = {};
    return Object.assign(base, extras);
}
function internalTestPropBase(testFn) {
    function base(label, arbitraries, prop, params) {
        buildTestWithPropRunner(testFn, label, arbitraries, prop, params, undefined, jest, fc);
    }
    const extras = {
        failing: internalTestPropFailing(testFn.failing),
    };
    return Object.assign(base, extras);
}
function internalTestPropConcurrent(testFn) {
    function base(label, arbitraries, prop, params) {
        buildTestWithPropRunner(testFn, label, arbitraries, prop, params, undefined, jest, fc);
    }
    const extras = {
        only: internalTestPropBase(testFn.only),
        skip: internalTestPropBase(testFn.skip),
        failing: internalTestPropFailing(testFn.failing),
    };
    return Object.assign(base, extras);
}
function internalTestProp(testFn) {
    const base = internalTestPropConcurrent(testFn);
    const extras = {
        concurrent: internalTestPropConcurrent(testFn.concurrent),
        todo: testFn.todo,
    };
    return Object.assign(base, extras);
}
export const buildTestProp = internalTestProp;
